lib/src/default_index/revset_engine.rs

```diff
@@ -916,15 +916,21 @@ impl EvaluationContext<'_> {
                     }
                 }
 
-                // Identify disjoint sets reachable from sources.
-                let set_reps: HashSet<_> = intersection_by(
-                    self.evaluate(sources)?.positions(),
-                    EagerRevWalk::new(domain_vec.iter().copied().map(Ok)),
-                    |pos1, pos2| pos1.cmp(pos2).reverse(),
-                )
-                .attach(index)
-                .map_ok(|pos| sets.find(pos))
-                .try_collect()?;
+                // Identify disjoint sets reachable from sources. Using a predicate here can be
+                // significantly faster for cases like `reachable(filter, X)`, since the filter
+                // can be checked for only commits in `X` instead of for all visible commits,
+                // and the difference is usually negligible for non-filter revsets.
+                let sources_revset = self.evaluate(sources)?;
+                let mut sources_predicate = sources_revset.to_predicate_fn();
+                let set_reps: HashSet<_> = domain_vec
```

Comment:
It seems like it should be possible to iterate over the sets and then include the whole set as soon as we find a single matching in the set, right? But it looks like UnionFind doesn't have an API for that yet. What you have looks like an improvement to me either way. I'll let Yuya approve since he knows the revset engine better.

Reply:
> we have to evaluate the predicate in order by reverse index position

Can you add a test that would otherwise break? I think some ancestry query should do the job.